### PR TITLE
feat(bigtable): Add Bind function to bind query params

### PR DESCRIPTION
**Pull request by bhshkh**
To execute a SQL query, user needs to 
1) **Prepare** query
The "Prepare" step takes a SQL query text and some additional information and returns some state and an execution plan. Essentially, it processes the query and sets up what is needed to run it, before the actual execution of the query occurs.
2) **Bind** parameters
3) **Execute** the query

This PR adds support for the second step.

More context:
- Related protos are here: [request](https://github.com/googleapis/google-cloud-go/blob/b323f888e349168b4e30a9edb4630dfb6060ed4d/bigtable/apiv2/bigtablepb/bigtable.pb.go#L1927-L1964), [response](https://github.com/googleapis/google-cloud-go/blob/b323f888e349168b4e30a9edb4630dfb6060ed4d/bigtable/apiv2/bigtablepb/bigtable.pb.go#L2050-L2066), [API](https://github.com/googleapis/google-cloud-go/blob/b323f888e349168b4e30a9edb4630dfb6060ed4d/bigtable/apiv2/bigtablepb/bigtable.pb.go#L2050-L2066)
- Design: go/cfdb-cbt-go-googlesql
- Previous BTQL PRs:
   https://github.com/googleapis/google-cloud-go/pull/11816
https://github.com/googleapis/google-cloud-go/pull/11829

### CODE REVIEW

File: bigtable/bigtable_test.go [hunk 1]
```diff
@@ -1118,3 +1122,528 @@ func TestApplyBulk_MutationsSucceedAfterGroupError(t *testing.T) {
...
+							},
+						},
+					},
+				},
+				Kind: &btpb.Value_ArrayValue{
+					ArrayValue: &btpb.ArrayValue{
+						Values: []*btpb.Value{
+							{
+								Type: &btpb.Type{
+									Kind: &btpb.Type_Int64Type{
```
Comment (jackdingilian) [inline bigtable/bigtable_test.go:1424]: Kind shouldn't be set in nested Values. It should only be at the top level

File: bigtable/bigtable.go [hunk 2]
```diff
@@ -54,6 +55,8 @@ const prodAddr = "bigtable.googleapis.com:443"
 const mtlsProdAddr = "bigtable.mtls.googleapis.com:443"
 const featureFlagsHeaderKey = "bigtable-features"
 
+var int64ReflectType = reflect.TypeOf(int64(0))
```
Comment (hongalex) [inline bigtable/bigtable.go:58]: consider moving to `sql_types.go`

**Review by hongalex — APPROVED** [review]

File: bigtable/bigtable_test.go (hunk 1 above)
Comment (jackdingilian) [reply bigtable/bigtable_test.go:1424]: Sorry this should have said 'Type' not 'Kind'

File: bigtable/bigtable.go (hunk 2 above)
Comment (bhshkh) [reply bigtable/bigtable.go]: Moved

File: bigtable/bigtable.go [hunk 3]
```diff
@@ -451,10 +460,52 @@ func (c *Client) prepareStatement(ctx context.Context, mt *builtinMetricsTracer,
...
+type BoundStatement struct {
+	ps     *PreparedStatement
+	params map[string]*btpb.Value
+}
+
+// Bind binds a set of parameters to a prepared statement.
+//
+// Allowed parameter value types are []byte, string, int64, float32, float64, bool,
+// time.Time, civil.Date, array, slice and nil
+func (ps *PreparedStatement) Bind(values map[string]any) (*BoundStatement, error) {
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply bigtable/bigtable.go:475]: Added validation and unit test

File: bigtable/bigtable_test.go (hunk 1 above)
Comment (bhshkh) [reply bigtable/bigtable_test.go]: removed

Comment (bhshkh) [reply bigtable/bigtable_test.go]: Updated

File: bigtable/sql_types.go [hunk 4]
```diff
@@ -18,20 +18,69 @@ package bigtable
...
+var int64ReflectType = reflect.TypeOf(int64(0))
+
 // SQLType represents the type of data that can be used to query Cloud Bigtable.
 // It is based on the GoogleSQL standard.
 type SQLType interface {
 	// Used while preparing the query
 	typeProto() (*btpb.Type, error)
+
+	// Used while binding parameters to prepared query
+	pbVal() *btpb.Value
```
Comment (gkevinzheng) [inline bigtable/sql_types.go:39]: For consistency, I think you should rename this function `valueProto` to be in line with the naming of `typeProto`.

File: bigtable/bigtable.go [hunk 5]
```diff
@@ -451,10 +457,59 @@ func (c *Client) prepareStatement(ctx context.Context, mt *builtinMetricsTracer,
...
+	bs := BoundStatement{
+		ps:     ps,
+		params: map[string]*btpb.Value{},
+	}
+	if values == nil {
+		return &bs, nil
+	}
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply bigtable/bigtable.go]: Handled

File: bigtable/sql_types.go [hunk 6]
```diff
@@ -141,3 +492,76 @@ func (s ArraySQLType) typeProto() (*btpb.Type, error) {
...
 }
+
+func anySQLTypeToPbVal(value any, sqlType SQLType) (*btpb.Value, error) {
+	switch t := sqlType.(type) {
+	case BytesSQLType:
+		sqlTypeVal, err := newBytesSQLType(value)
+		if err != nil {
+			return nil, err
+		}
+		return sqlTypeVal.pbVal(), nil
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply bigtable/sql_types.go]: Refactored

File: bigtable/sql_types.go (hunk 4 above)
Comment (bhshkh) [reply bigtable/sql_types.go]: renamed

**Review by gkevinzheng — APPROVED** [review]

File: bigtable/bigtable.go [hunk 7]
```diff
@@ -396,6 +396,12 @@ type PrepareOption interface {
 
 // PrepareStatement prepares a query for execution. If possible, this should be called once and
 // reused across requests. This will amortize the cost of query preparation.
+//
+// The query string can be a parameterized query containing placeholders in the form of @ followed by the parameter name
+// Parameter names may consist of any combination of letters, numbers, and underscores.
+//
+// Parameters can appear anywhere that a literal value is expected. The same parameter name can
+// be used more than once, for example: WHERE cf["qualifier1"] = @value OR cf["qualifier2"] * = @value
```
Comment (jackdingilian) [inline bigtable/bigtable.go:404]: `OR cf["qualifier2"] * = @value` I think the `*` character should be removed

File: bigtable/bigtable_test.go [hunk 8]
```diff
@@ -1118,3 +1122,526 @@ func TestApplyBulk_MutationsSucceedAfterGroupError(t *testing.T) {
...
+				Type: &btpb.Type{
+					Kind: &btpb.Type_Float32Type{
+						Float32Type: &btpb.Type_Float32{},
+					},
+				},
+			},
+		},
+		{
+			testName:   "Float32SQLType type mismatch",
+			paramVal:   "1.23",
```
Comment (jackdingilian) [inline bigtable/bigtable_test.go:1265]: nit: maybe we should test this w float64? That feels like the most likely type mismatch here. And same for the reverse below

File: bigtable/sql_types.go [hunk 9]
```diff
@@ -18,20 +18,69 @@ package bigtable
...
+	// Used while binding parameters to prepared query
+	valueProto() *btpb.Value
+
+	isValidArrayElemType() bool
 }
 
 // BytesSQLType represents a slice of bytes.
-type BytesSQLType struct{}
+type BytesSQLType struct {
+	value *btpb.Value
```
Comment (jackdingilian) [inline bigtable/sql_types.go:46]: Does the type need to be stateful like this? It feels wrong to me for a type to hold a value. 

Could we make it so that valueProto is something like  toValueProto(value any) *btpb.Value?

Comment (bhshkh) [reply bigtable/sql_types.go]: Updated

File: bigtable/bigtable_test.go (hunk 8 above)
Comment (bhshkh) [reply bigtable/bigtable_test.go:1265]: Added scenario

File: bigtable/bigtable.go (hunk 7 above)
Comment (bhshkh) [reply bigtable/bigtable.go]: Updated

File: bigtable/sql_types.go [hunk 10]
```diff
@@ -42,8 +82,36 @@ func (s BytesSQLType) typeProto() (*btpb.Type, error) {
 
 // StringSQLType represents a string.
 type StringSQLType struct {
+	value *btpb.Value
```
Comment (jackdingilian) [inline bigtable/sql_types.go:85]: This can be removed now right? For the other types as well

File: bigtable/bigtable_test.go [hunk 11]
```diff
@@ -1118,3 +1122,540 @@ func TestApplyBulk_MutationsSucceedAfterGroupError(t *testing.T) {
...
+									IntValue: int64(3),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			testName: "ArraySQLType nil success",
```
Comment (jackdingilian) [inline bigtable/bigtable_test.go:1537]: Can you also add an empty array test, to make sure it's treated distinctly from null array

**Review by jackdingilian — APPROVED** [review]

File: bigtable/sql_types.go (hunk 10 above)
Comment (bhshkh) [reply bigtable/sql_types.go]: Removed

File: bigtable/bigtable_test.go (hunk 11 above)
Comment (bhshkh) [reply bigtable/bigtable_test.go:1535]: Added
